homeassistant/components/climate/nuheat.py

```diff
@@ -0,0 +1,184 @@
+    api, serial_numbers = hass.data[DATA_NUHEAT]
+    thermostats = [
+        NuHeatThermostat(api, serial_number, temperature_unit)
+        for serial_number in serial_numbers
+    ]
+    add_devices(thermostats, True)
+
+
+class NuHeatThermostat(ClimateDevice):
+    """Representation of a NuHeat Thermostat."""
+
+    def __init__(self, api, serial_number, temperature_unit):
+        """Initialize the thermostat."""
+        self._thermostat = api.get_thermostat(serial_number)
+        self._temperature_unit = temperature_unit
+        self._force_update = False
+
+    @property
+    def name(self):
+        """Return the name of the thermostat."""
+        return self._thermostat.room
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend."""
+        return ICON
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return SUPPORT_FLAGS
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        if self._temperature_unit == "C":
+            return TEMP_CELSIUS
+
+        return TEMP_FAHRENHEIT
+
+    @property
+    def current_temperature(self):
+        """Return the current temperature."""
+        if self._temperature_unit == "C":
+            return self._thermostat.celsius
+
+        return self._thermostat.fahrenheit
+
+    @property
+    def current_operation(self):
+        """Return current operation. ie. heat, idle."""
+        if self._thermostat.heating:
+            return STATE_HEAT
+
+        return STATE_IDLE
+
+    @property
+    def min_temp(self):
+        """Return the minimum supported temperature for the thermostat."""
+        if self._temperature_unit == "C":
+            return self._thermostat.min_celsius
+
+        return self._thermostat.min_fahrenheit
+
+    @property
+    def max_temp(self):
+        """Return the maximum supported temperature for the thermostat."""
+        if self._temperature_unit == "C":
+            return self._thermostat.max_celsius
+
+        return self._thermostat.max_fahrenheit
+
+    @property
+    def target_temperature(self):
+        """Return the currently programmed temperature."""
+        if self._temperature_unit == "C":
+            return self._thermostat.target_celsius
+
+        return self._thermostat.target_fahrenheit
+
+    @property
+    def current_hold_mode(self):
+        """Return current hold mode."""
+        schedule_mode = self._thermostat.schedule_mode
+        if schedule_mode == SCHEDULE_RUN:
+            return MODE_AUTO
+
+        if schedule_mode == SCHEDULE_HOLD:
+            return MODE_HOLD_TEMPERATURE
+
+        if schedule_mode == SCHEDULE_TEMPORARY_HOLD:
+            return MODE_TEMPORARY_HOLD
+
+        return MODE_AUTO
+
+    @property
+    def operation_list(self):
+        """Return list of possible operation modes."""
+        return OPERATION_LIST
+
+    def resume_program(self):
```

Comment:
When is this used?

Reply:
The thermostat has a schedule programmed into it. If you would override a temperature with HASS, the programmed schedule would be paused. To return back to the scheduled program, you would call `resume_program`

For example, if you left the house, HASS could set the temperature very low to prevent the floor from heating. Upon your return, HASS would tell the thermostat to resume its scheduled program.

Reply:
But there's no service for `resume_program`.

Reply:
What do you mean? It's defined right here and tested below.

Reply:
Every entity method that is not only used internally needs to have a corresponding service defined from which the user can call the method via the home assistant api. How would this method be called otherwise?

Reply:
ahhh, i see. you are saying there's no service defined on the general `climate` component. i'll look into how other thermostat components revert back to a thermostat's schedule.

Reply:
Wait, the ecobee component has a `resume_program` service almost identical to this.
Since these devices extend the `ClimateDevice`, that service should just be available, right?

Reply:
No that service is only defined in the ecobee platform, so it's only available to that platform. If we think this is a general enough feature for climate devices we could consider adding the service to the climate component. @tinloaf what do you think?

Reply:
oh, i see that they also register the service with `hass.services.register`. that is something i can easily add here.

otherwise, it seems that we would use the `set_operation_mode` with the `STATE_AUTO` flag ... however, i feel like that's just going to re-open up that whole operation_mode vs hold_mode debate, heh.

Reply:
I'm not sure we should have a separate service for this. I think "the device follows its pre-programmed schedule" is an operation mode. So, setting the `operation_mode` to `AUTO` (or whatever) should do what calling `resume_schedule` does, right? If changing some setting in HA should interrupt this behavior, it should also change the `operation_mode` to `MANUAL` or something, I think. 

Adding this as a separate service adds another layer of "setting how the device controls its temperature", besides `operation_mode`, `hold_mode` and `away` - which I think are already way too many.

Reply:
P.S.: I think to be consistent with the behavior of most other platforms, manually setting a temperature should *not* set the `operation_mode` to `MANUAL`. For most platforms (I think), the behavior is this:
* While in `AUTO` mode, manually setting the temperature takes effect until the next control point of the `AUTO` mode is reached, i.e., the device still follows its pre-programmed schedule.
* While in `MANUAL` mode, the manually set temperature is held indefinitely.

Reply:
This is what I was worried about. 😄

@tinloaf, you're injecting your personal bias on how you believe operation modes and hold modes should work, rather than how they're currently documented and implemented. I'm with you in that I think there's room for improvement on the HASS climate component modes... but I'm just following the current standards, docs, and implementations within HASS.

Based on the documentation here and on NuHeat, the only effective **operation modes** for this thermostat are: `heat` and `idle`. They are not directly controllable, but are implicitly toggled by setting the target temperature higher or lower than the current temperature.

The **hold modes** for this thermostat are: auto (run schedule programmed into thermostat), indefinite hold at a target temperature, and temporary hold at a target temperature. They are explicitly called hold modes on the NuHeat website, thermostat, and web UI and match up with the documentation here.

The thermostat itself and their web UIs both have "Resume program" buttons that appear when in a temporary or indefinite hold. As such, I've implemented a simple `nuheat_resume_program` service to be consistent with that.

Regarding your comment on how there are way too many services for setting how the device controls its temperature... There should only be 1 for this thermostat: Hold mode. The operation modes are not toggle-able and I've removed the `home` and `away` modes at your request.

If this component is not helpful or too controversial, I can just close this pull request and implement a custom component for my own use.

Reply:
Hi @broox - first things first: Thanks for all your work, and for putting up with me. I know I can be perfectionist some times, but I'd also love to see this PR merged. However, when reviewing PRs, I try to comment on how things *should* be (and of course that's some kind of opinion then), not how they *are* currently. Nonetheless I think that for most current platforms, *operation* modes are mostly used for what you said would be *hold* modes. I completely agree that it is not clear what the exact distinction between these two should be and that we should clean this up.

This will hopefully all be cleaned up soon. Thus, I don't think there is too much damage done in getting such a change in (even If I might think it's not consistent :wink: ). However, I would ask you not to be offended if we might change this behavior once we do the cleanup.

Reply:
That sounds totally fair. I'm happy to refactor, review and/or test this component once things are cleaned up more generally.